### PR TITLE
Prepare FEVER wiki dumps for indexing with anserini

### PR DESCRIPTION
**Pull request by kelvin-jiang**
Changes:
* Added a script that converts FEVER dataset (pre-processed Wikipedia introductions dumps) into JSONL format to allow for indexing with anserini

Some outstanding questions (@ronakice):
1. FEVER training data consists of claims with evidence to support/refute attached to each claim. The attached evidence are on a sentence-level, but currently I have it so that the FEVER dataset is indexed on a document (Wikipedia introduction) level. Should we be indexing on a sentence-level?
2. The FEVER training data refers to evidence by their Wikipedia article ID. Do we need a script that converts these Wikipedia article IDs to their new doc IDs that are in the converted JSONL files?
3. The FEVER wiki dumps have been pre-processed, and it looks like the sentences have been tokenized using some sort of Penn Treebank tokenization (ftp://ftp.cis.upenn.edu/pub/treebank/public_html/tokenization.html). Do we need to detokenize these sentences (in case it impacts the indexing/training somehow)?

Commands to prepare, index, and retrieve on dataset (putting it here for myself, but also for anyone else to reproduce):
```
wget https://s3-eu-west-1.amazonaws.com/fever.public/wiki-pages.zip -P collections/fever

unzip collections/fever/wiki-pages.zip -d collections/fever

python3 src/main/python/fever/convert_collection_to_jsonl.py --collection_path collections/fever/wiki-pages --output_folder collections/fever/wiki-pages-jsonl

sh target/appassembler/bin/IndexCollection -collection JsonCollection -generator DefaultLuceneDocumentGenerator -threads 9 -input collections/fever/wiki-pages-jsonl -index indexes/fever/lucene-index-fever -storePositions -storeDocvectors -storeRaw

python3 src/main/python/msmarco/retrieve.py --hits 100 --threads 1 --index indexes/fever/lucene-index-fever --qid_queries collections/fever/queries.dev.small.tsv --output runs/run.fever.dev.small.tsv
```

### CODE REVIEW

**Comment by lintool** [comment]
Can we write an ingestor for the original source document so we can skip the conversion?

Custom formats aren't that hard to write... see: https://github.com/castorini/anserini/pull/1266

Data munging is error prone... may seem like shortcut now, but you lose track of what version of the munging script is associate with which index pretty quickly... reproducibility is a pain.

**Comment by kelvin-jiang** [comment]
Initially, @ronakice and I decided that the dataset was too specific to warrant dedicated classes, but yeah I can build those out. I'll wait for my above questions to get addressed before I work on that.

**Comment by lintool** [comment]
> * FEVER training data consists of claims with evidence to support/refute attached to each claim. The attached evidence are on a sentence-level, but currently I have it so that the FEVER dataset is indexed on a document (Wikipedia introduction) level. Should we be indexing on a sentence-level?

Both? In the same way the basic CORD-19 classes support traversing documents at multiple granularities (abstract, full-text, paragraph), we could do the same here.

> * The FEVER training data refers to evidence by their Wikipedia article ID. Do we need a script that converts these Wikipedia article IDs to their new doc IDs that are in the converted JSONL files?

Probably? I'll let @ronakice answer this?

> * The FEVER wiki dumps have been pre-processed, and it looks like the sentences have been tokenized using some sort of Penn Treebank tokenization (ftp://ftp.cis.upenn.edu/pub/treebank/public_html/tokenization.html). Do we need to detokenize these sentences (in case it impacts the indexing/training somehow)?

I wouldn't worry about it in the first pass... let's get something going and see how much poor tokenziation is hurting...

**Comment by ronakice** [comment]
> * FEVER training data consists of claims with evidence to support/refute attached to each claim. The attached evidence are on a sentence-level, but currently I have it so that the FEVER dataset is indexed on a document (Wikipedia introduction) level. Should we be indexing on a sentence-level?
Like what @lintool said, let's go with both levels of granularities.

> * The FEVER training data refers to evidence by their Wikipedia article ID. Do we need a script that converts these Wikipedia article IDs to their new doc IDs that are in the converted JSONL files?

We probably should.

> * The FEVER wiki dumps have been pre-processed, and it looks like the sentences have been tokenized using some sort of Penn Treebank tokenization (ftp://ftp.cis.upenn.edu/pub/treebank/public_html/tokenization.html). Do we need to detokenize these sentences (in case it impacts the indexing/training somehow)?

Let's just go with tokenization as is. It might add "length" to the tokenization input but for the time being doesn't matter too much

Where's `collections/fever/queries.dev.small.tsv` btw? I'm trying to replicate, will then get back to you.

**Comment by kelvin-jiang** [comment]
Oh, I just manually put some random queries from the FEVER training data into `collections/fever/queries.dev.small.tsv` for testing purposes. I can convert all of the training data into a TSV file if you want.

**Comment by ronakice** [comment]
Yup, let's also do that for train/dev/test (whatever they provide!)

**Comment by kelvin-jiang** [comment]
New changes:
* Created `FeverParagraphCollection.py` and `FeverSentenceCollection.py` that ingests the FEVER wikipedia dumps by their respective granularity
  * To index, use `DefaultLuceneDocumentGenerator` as `generator` argument
* Added `generate_queries_and_qrels.py` that converts FEVER datasets (train/test/dev) to TREC queries and qrels format
  * Download dataset splits:
```
wget https://s3-eu-west-1.amazonaws.com/fever.public/train.jsonl -P collections/fever
wget https://s3-eu-west-1.amazonaws.com/fever.public/shared_task_dev.jsonl -P collections/fever
wget https://s3-eu-west-1.amazonaws.com/fever.public/shared_task_test.jsonl -P collections/fever
```

Also updated the original `convert_collection_to_jsonl.py` script to support both granularities, which allowed me to test my ingestors (e.g. indexed with FeverParagraphCollection -> indexed with JsonCollection after running the conversion script -> verified retrieval results to be the same with both indexes).

**Comment by kelvin-jiang** [comment]
FTR @lintool, I fixed the weird NPE bug with this check: https://github.com/castorini/anserini/pull/1271/files#diff-58db00bfebca71d9041fe72405732dd8R79.

For some reason, even though I inserted this entry: `expected.put("", Map.of("id", "", "content", "", "raw", ""));`, `expected.get("")` which is called by `DocumentCollectionTest` would return null instead of the map I inserted. Therefore, when I tried to get "id" or "content" from that map during my unit tests, it would raise an NPE.

**Comment by lintool** [comment]
hey @kelvin-jiang I'm lacking the broader context here, so I'll let @ronakice take the first crack?

File: src/test/resources/sample_docs/fever/collection1/segment1.jsonl [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{"id": "", "text": "", "lines": ""}
+{"id": "Domain_Range_Ratio_-LRB-DRR-RRB-", "text": "", "lines": ""}
+{"id": "Shohei_Otani", "text": "is a Japanese baseball pitcher , outfielder , and designated hitter who plays for the Hokkaido Nippon-Ham Fighters of Nippon Professional Baseball 's -LRB- NPB -RRB- Pacific League . He was the first pick of the Fighters in the 2012 draft . He officially recorded the fastest pitch by a Japanese pitcher and in NPB history at 165 km/h . ", "lines": "0\tis a Japanese baseball pitcher , outfielder , and designated hitter who plays for the Hokkaido Nippon-Ham Fighters of Nippon Professional Baseball 's -LRB- NPB -RRB- Pacific League .\tHokkaido Nippon-Ham Fighters\tHokkaido Nippon-Ham Fighters\tbaseball\tbaseball\tpitcher\tpitcher\tNippon Professional Baseball\tNippon Professional Baseball\tPacific League\tPacific League\n1\tHe was the first pick of the Fighters in the 2012 draft .\n2\tHe officially recorded the fastest pitch by a Japanese pitcher and in NPB history at 165 km/h .\tpitcher\tpitcher\n3\t"}
```
Comment (ronakice) [inline src/test/resources/sample_docs/fever/collection1/segment1.jsonl:3]: Is the first line always like this?

File: src/test/java/io/anserini/collection/FeverSentenceCollectionTest.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+                    "Dakota .",
+            "raw", "Kelvin is an unincorporated community in Rolette County , in the U.S. state of North " +
+                    "Dakota ."
+    ));
+    expected.put("Kelvin,_North_Dakota_1", Map.of("id", "Kelvin,_North_Dakota_1", "content", "", "raw", ""));
+
+    // regular document 3 with unicode characters and empty lines, should be ingested and indexed
+    expected.put("Styrkepr\u00f8ven_0", Map.of(
+            "id", "Styrkepr\u00f8ven_0",
+            "content", "Styrkepr\u00f8ven , also called Den Store Styrkepr\u00f8ven -LRB- The Great Trial of " +
```
Comment (ronakice) [inline src/test/java/io/anserini/collection/FeverSentenceCollectionTest.java:85]: Should we use ftfy to fix encoding here? I think it works?

`fix_encoding("Styrkepr\u00f8ven")
`
Styrkeprøven

File: src/test/resources/sample_docs/fever/collection1/segment1.jsonl (hunk 1 above)
Comment (kelvin-jiang) [reply src/test/resources/sample_docs/fever/collection1/segment1.jsonl:3]: Nope, this was just an edge case I found in the wikipedia dumps.

File: src/test/java/io/anserini/collection/FeverSentenceCollectionTest.java (hunk 2 above)
Comment (kelvin-jiang) [reply src/test/java/io/anserini/collection/FeverSentenceCollectionTest.java:85]: I'm not 100% sure, but I thought `ftfy` was used to fix broken unicode? This is from their website:
> The word schön might appear as schÃ¶n

Strings in python3 are unicode-encoded by default, so I think we should be able to get desired results for free:
```
>>> print('Styrkepr\u00f8ven')
Styrkeprøven
```

File: src/main/python/fever/convert_collection_to_jsonl.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
+# -*- coding: utf-8 -*-
```
Comment (lintool) [inline src/main/python/fever/convert_collection_to_jsonl.py:1]: We've tweaked our header, please use this? https://github.com/castorini/anserini-tools/blob/master/scripts/msmarco/convert_msmarco_to_trec_qrels.py#L1

Also, we recently learned that according to PEP8, the files _shouldn't_ have coding directive.

File: src/main/python/fever/generate_queries_and_qrels.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
+# -*- coding: utf-8 -*-
```
Comment (lintool) [inline src/main/python/fever/generate_queries_and_qrels.py:1]: Same here.

**Comment by lintool** [comment]
Hi @kelvin-jiang - sorry about the delay, just caught up on this. Minor changes, but otherwise ready for merging.

I assume you'll write up a IR baseline replication guide at some point in time? Please do.

File: src/main/python/fever/convert_collection_to_jsonl.py (hunk 3 above)
Comment (kelvin-jiang) [reply src/main/python/fever/convert_collection_to_jsonl.py]: Updated!

File: src/main/python/fever/generate_queries_and_qrels.py (hunk 4 above)
Comment (kelvin-jiang) [reply src/main/python/fever/generate_queries_and_qrels.py]: Updated as well.

**Comment by kelvin-jiang** [comment]
@lintool For sure, I can write up that replication guide. I'll probably include that in a separate PR along with the work for comparing anserini results with the FEVER paper (as discussed on Slack) so this PR doesn't get too cluttered.

**Comment by ronakice** [comment]
Awesome! Let's get the Table 2 results first plus the T5 input files. Then you can add the replication guide!

**Comment by lintool** [comment]
Nice thanks for your contribution @kelvin-jiang !
